filters/encrypt/filter.go

```diff
@@ -0,0 +1,542 @@
+		ef.HmacSalt = opts.withSalt
+	}
+	if opts.withInfo != nil {
+		ef.HmacInfo = opts.withInfo
+	}
+}
+
+// Process will encrypt or hmac-sha256 string and []byte fields which are tagged
+// as SensitiveClassification.  Fields that are tagged SecretClassification will
+// be redacted.
+//
+// If the event payload satisfies the WrapperPayload interface, then the
+// payload's Wrapper(), HmacSalt() and HmacInfo() will be used to rotate the
+// filter's wrappers for ongoing filtering operations.  Events matching this
+// WrapperPayload interface are not sent along in the pipeline and a nil with no
+// errors is immediately returned after the wrapper has been rotated.
+//
+// If the event payload satisfies the EventWrapperInfo interface, then the
+// payload's EventId(), HmacSalt() and HmacInfo() will be used to for filtering
+// operations for just the single event being processed.
+func (ef *Filter) Process(ctx context.Context, e *eventlogger.Event) (*eventlogger.Event, error) {
+	const op = "event.(Filter).Process"
+	if e == nil {
+		return nil, fmt.Errorf("%s: missing event: %w", op, ErrInvalidParameter)
+	}
+	if i, ok := e.Payload.(RotateWrapper); ok {
+		ef.l.Lock()
+		defer ef.l.Unlock()
+		if i.Wrapper() != nil {
+			ef.Wrapper = i.Wrapper()
+		}
+		if i.HmacSalt() != nil {
+			ef.HmacSalt = make([]byte, len(i.HmacSalt()))
+			copy(ef.HmacSalt, i.HmacSalt())
+		}
+		if i.HmacInfo() != nil {
+			ef.HmacInfo = make([]byte, len(i.HmacInfo()))
+			copy(ef.HmacInfo, i.HmacInfo())
+		}
+		return nil, nil
+	}
+
+	var opts []Option
+	var optWrapper wrapping.Wrapper
+	if i, ok := e.Payload.(EventWrapperInfo); ok {
+		ef.l.RLock()
+		w, err := NewEventWrapper(ef.Wrapper, i.EventId())
+		ef.l.RUnlock()
+		if err != nil {
+			return nil, fmt.Errorf("%s: %w", op, err)
+		}
+		optWrapper := w
+		opts = append(opts, WithWrapper(optWrapper))
+		opts = append(opts, WithInfo(i.HmacInfo()))
+		opts = append(opts, WithSalt(i.HmacSalt()))
+	}
+
+	if ef.Wrapper == nil && optWrapper == nil {
+		return nil, fmt.Errorf("%s: missing wrapper: %w", op, ErrInvalidParameter)
+	}
+
+	if e.Payload == nil {
+		return e, nil
+	}
+
+	// Get both the value and the type of what the payload points to. Value is
+	// used to mutate underlying data and Type is used to get the name of the
+	// field.
+	payloadValue := reflect.ValueOf(e.Payload)
+	switch payloadValue.Kind() {
+	case reflect.Ptr, reflect.Interface:
+		payloadValue = reflect.ValueOf(e.Payload).Elem()
+	}
+
+	pType := payloadValue.Type()
+	pKind := payloadValue.Kind()
+
+	taggedInterface, isTaggable := payloadValue.Interface().(Taggable)
+
+	switch {
+	case pType == reflect.TypeOf("") || pType == reflect.TypeOf([]uint8{}):
+		if !payloadValue.CanSet() {
+			return nil, fmt.Errorf("%s: unable to redact string payload (not setable): %w", op, ErrInvalidParameter)
+		}
+		ef.l.RLock()
+		classificationTag := getClassificationFromTagString(string(SecretClassification), withFilterOperations(ef.FilterOperationOverrides))
```

Comment:
It seems possible that this map could change partway through processing fields. Additionally it doesn't seem worthwhile to rlock on (potentially) every field. Can you just get the FilterOperationOverrides value at the top with a single rlock?

Reply:
ty!  I've made a copy of the override set at the top which will give us a consistent set while processing the event.